pkg/proc/core/minidump/minidump.go

```diff
@@ -0,0 +1,689 @@
+		case MiscInfoStream:
+			readMiscInfo(&mdmp, streamBuf(stream, &buf, "misc info"))
+			if logfn != nil {
+				logfn("\tPid: %#x\n", mdmp.Pid)
+			}
+		case CommentStreamW:
+			if logfn != nil {
+				logfn("\t%q\n", decodeUTF16(stream.RawData))
+			}
+		case CommentStreamA:
+			if logfn != nil {
+				logfn("\t%s\n", string(stream.RawData))
+			}
+		}
+		if buf.err != nil {
+			return nil, buf.err
+		}
+	}
+
+	return &mdmp, nil
+}
+
+// decodeUTF16 converts a NUL-terminated UTF16LE string to (non NUL-terminated) UTF8.
+func decodeUTF16(in []byte) string {
+	utf16encoded := []uint16{}
+	for i := 0; i+1 < len(in); i += 2 {
+		var ch uint16
+		ch = uint16(in[i]) + uint16(in[i+1])<<8
+		utf16encoded = append(utf16encoded, ch)
+	}
+	s := string(utf16.Decode(utf16encoded))
+	if len(s) > 0 && s[len(s)-1] == 0 {
+		s = s[:len(s)-1]
+	}
+	return s
+}
+
+func fileFlagsToString(flags FileFlags) string {
+	out := []byte{}
+	for i, name := range _FileFlags_map {
+		if i == 0 {
+			continue
+		}
+		if flags&i != 0 {
+			if len(out) > 0 {
+				out = append(out, '|')
+			}
+			out = append(out, name...)
+		}
+	}
+	if len(out) == 0 {
+		return flags.String()
+	}
+	return string(out)
+}
+
+// readMinidumpHeader reads the minidump file header
+func readMinidumpHeader(mdmp *Minidump, buf *minidumpBuf) {
+	buf.ctx = "reading minidump header"
+
+	if sig := buf.u32(); sig != minidumpSignature {
+		buf.err = ErrNotAMinidump{"signature", sig}
+		return
+	}
+
+	if ver := buf.u16(); ver != minidumpVersion {
+		buf.err = ErrNotAMinidump{"version", uint32(ver)}
+		return
+	}
+
+	buf.u16() // implementation specific version
+	mdmp.streamNum = buf.u32()
+	mdmp.streamOff = buf.u32()
+	buf.u32() // checksum, but it's always 0
+	mdmp.Timestamp = buf.u32()
+	mdmp.Flags = FileFlags(buf.u64())
+}
+
+// readDirectory reads the list of streams (i.e. the minidum "directory")
+func readDirectory(mdmp *Minidump, buf *minidumpBuf) {
+	buf.off = int(mdmp.streamOff)
+
+	mdmp.Streams = make([]Stream, mdmp.streamNum)
+	for i := range mdmp.Streams {
+		buf.ctx = fmt.Sprintf("reading stream directory entry %d", i)
+		stream := &mdmp.Streams[i]
+		stream.Type = StreamType(buf.u32())
+		stream.Offset, stream.RawData = readLocationDescriptor(buf)
+		if buf.err != nil {
+			return
+		}
+	}
+}
+
+// readLocationDescriptor reads a location descriptor structure (a structure
+// which describes a subregion of the file), and returns the destination
+// offset and a slice into the minidump file's buffer.
+func readLocationDescriptor(buf *minidumpBuf) (off int, rawData []byte) {
+	sz := buf.u32()
+	off = int(buf.u32())
+	if buf.err != nil {
+		return off, nil
+	}
+	end := off + int(sz)
+	if off >= len(buf.buf) || end > len(buf.buf) {
+		buf.err = fmt.Errorf("location starting at %#x of size %#x is past the end of file, while %s", off, sz, buf.ctx)
+		return 0, nil
+	}
+	rawData = buf.buf[off:end]
+	return
+}
+
+func readString(buf *minidumpBuf) string {
```

Comment:
Could we just reuse `pkg/dwarf/util.ParseString` here?

Reply:
ParseString reads a NUL terminated string, this is a pascal-like string.